Cleanup duplicates in MappingMethods

filiphr opened the pull request:
I have created a new class for all the `MappingMethods` that have a factory method. There was a lot of duplication between all of the different methods.

Is it too much abstraction (with the new class)?

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/model/MappingMethodWithFactory.java:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @author Filip Hrisafov
+ */
+public abstract class MappingMethodWithFactory extends MappingMethod {
```
Is it the fact that it has a factory that sets apart methods inheriting from this method? I reckon in essence these are generated non-enum / non-value mapping methods: MapMapping, BeanMapping, IterableMapping, StreamMapping right?

filiphr replied:
You are right, it is for those 4 Mappings. The factory was something they shared so I used that in the name. However, I am open for suggestions, I am sure it can be better named

sjaakd replied:
jls chapter 8.1 (https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html) makes a distinction between an 
```
ClassDeclaration:
    NormalClassDeclaration
    EnumDeclaration

```

so what about `NormalTypeMapper`.

sjaakd replied:
excuse me : `NormalTypeMappingMethod`

filiphr replied:
That is interesting, I never knew this 😄. I'll rename it to `NormalTypeMappingMethod`.

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/ContainerMappingMethod.java:
```diff
@@ -36,11 +36,8 @@
  *
  * @author Filip Hrisafov
  */
-public abstract class ContainerMappingMethod extends MappingMethod {
```
You mean `loopVariableName`, right? This one is actually needed:
* In `IterableMappingMethod.ftl` - here it is clear
* In `StreamMappingMethod.ftl` -for updating an array on line 103

Both of them have some form of element assignment and use the `SelectionParameters`. Theoretically other similar mappings (for example Guava collections) can benefit from the `ContainerMappingMethod`.

As for `getDefault()`, it was in the `IterableMappingMethod` and I pulled it up when I needed it for the `StreamMappingMethod`. I didn't know about `Type#getNull()`, had I known I would have used it. I can clean that up as well and use the one from the `Type`. We will just have to agree on the differences between the 2.

sjaakd replied:
Ok.. but if the getDefault() moves what does remain here? And does it justify to have a common base class. I'm not in favor of long inheritance chains, and prefer composition of inheritance if possible. Anyway. Don't consider this as blocker.

filiphr replied:
This methods will remain:
* `getLoopVariableName`, `getIndex1Name` and `getIndex2Name` - for the arrays
* `getElementAssignment` - How the underlying element is assigned.
* `getResultElementType` - I think that this is the most important one. An abstract that needs to be implemented to get the `Type` of the result element. 
* The `selectionParameters`

I completely understand you not liking the long inheritance chain. Do you perhaps have some other suggestion? Maybe make the `MappingMethodWithFactory` (which will be renamed) generic and use that for the `MappingMethod` specifics?

sjaakd replied:
> I completely understand you not liking the long inheritance chain. Do you perhaps have some other suggestion?

I think we should park this one. But on thing we might consider is to use composition instead of inheritance. Or a combination: we still have a base class `MappingMethod` and top level methods like `BeanMappingMethod`, `MapMappingMethod` but we compose other behavior out of the `ContainerMappingMethod` and ` NormalTypeMappingMethod`

filiphr replied:
Should I do this as part of this PR or create a new issue to do it for later?

sjaakd replied:
Someday.  A new issue is fine, but not urgent. It works as is. That's what I meant with: 

> I think we should park this one.

filiphr replied:
OK. I created #1066 for it

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java:
```diff
@@ -758,10 +758,11 @@ public String getNull() {
         }
         if ( "double".equals( getName() ) ) {
-            return "0.0";
+            return "0.0d";
```
I alligned this one with the default value that was use before. I did the same for the `char`.

We also need to decide whether we want to do it like we did before with the `TypeMirror` `Kind` or like it is here.

sjaakd replied:
Does it still work for the other scenarios (where `Type` is used directly)?

filiphr replied:
Well nothing failed. The other scenario where this was used is the `NestedPropertyMappingMethod.ftl`.

filiphr commented:
@sjaakd I addressed your comments and renamed the `MappingMethodWithFactory`. I also am using the `Type#getNull()` methods instead of the `ContainerMappingMethod#getDefaultValue()`. I alligned the `getNull()` with how it was for the `getDefaultValue()`.

sjaakd commented on processor/src/main/resources/org/mapstruct/ap/internal/model/IterableMappingMethod.ftl:
```diff
@@ -35,7 +35,7 @@
                     <#-- we can't clear an existing array, so we've got to clear by setting values to default -->
                     for (int ${index2Name} = 0; ${index2Name} < ${resultName}.length; ${index2Name}++ ) {
-                        ${resultName}[${index2Name}] = ${defaultValue};
+                        ${resultName}[${index2Name}] = ${resultElementType.null};
```
👍

sjaakd replied:
resultElementType cannot be null?

filiphr replied:
I don't think so. Just have a look at the `getDefaultValue()` that was removed. We were doing `getResultElementType().getTypeMirror()...` so if it was `null` we would have had an NPE.

sjaakd commented:
Just a few questions remaining. If you feel ok, please proceed.

filiphr commented:
Thanks for the review @sjaakd
